### PR TITLE
fix: [SRM-16172]: Add/Refactor update validation, add index

### PR DESCRIPTION
**Pull request by naveen-harness**
## Describe your changes

Change sources and health sources are templatised while Notifications and dependencies aren't, based on that we've refactored the validation flow. The validation rules are given below for the monitored service reconciliation:
Screenshot 2023-11-24 at 3 44 20 PM

Additionaly, fixed an issue where we were using isTemplateByReference from request body, it should be fetched from the entity instead. 

Furthermore, added a compound mongo db index for faster query.

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/55938)

### CODE REVIEW

**Comment by naveen-harness** [comment]
trigger uts

File: srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/core/services/impl/monitoredService/MonitoredServiceServiceImpl.java [hunk 1]
```diff
@@ -535,12 +530,17 @@ private void updateMonitoredService(
                                                         .map(HealthSource::getIdentifier)
                                                         .collect(Collectors.toList());
       updateOperations.set(MonitoredServiceKeys.healthSourceIdentifiers, updatedHealthSourceIdentifiers);
-      List<String> updatedChangeSourceIdentifiers = monitoredServiceDTO.getSources()
-                                                        .getChangeSources()
-                                                        .stream()
-                                                        .map(ChangeSourceDTO::getIdentifier)
-                                                        .collect(Collectors.toList());
-      updateOperations.set(MonitoredServiceKeys.changeSourceIdentifiers, updatedChangeSourceIdentifiers);
+      // if the update request is coming from reconciliation flow, then
```
Comment (karansaraswat19) [inline srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/core/services/impl/monitoredService/MonitoredServiceServiceImpl.java:533]: will need to do this for healthSourceIdentifiers as well

File: srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/core/services/impl/monitoredService/MonitoredServiceServiceImpl.java [hunk 2]
```diff
@@ -556,19 +556,20 @@ private void updateMonitoredService(
     if (monitoredServiceDTO.getTemplate() != null
         && featureFlagService.isFeatureFlagEnabled(
             projectParams.getAccountIdentifier(), FeatureFlagNames.SRM_ENABLE_MS_TEMPLATE_RECONCILIATION)
-        && monitoredServiceDTO.getTemplate().isTemplateByReference() && isUpdatedFromYaml) {
+        && monitoredService.getTemplateMetadata().isTemplateByReference() && isUpdatedFromYaml) {
```
Comment (karansaraswat19) [inline srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/core/services/impl/monitoredService/MonitoredServiceServiceImpl.java:559]: we can directly use the isTemplateByReference method from the entity

File: srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/core/services/impl/monitoredService/MonitoredServiceServiceImpl.java [hunk 3]
```diff
@@ -495,31 +499,22 @@ public MonitoredServiceResponse update(
...
-            String.format("Update of Monitored Service entity with identifier: %s is not allowed"
-                    + "with notification rules if it's created through template by reference.",
-                monitoredServiceDTO.getIdentifier()));
-      }
-      if (isNotEmpty(monitoredServiceDTO.getDependencies())) {
-        throw new InvalidRequestException(
-            String.format("Update of Monitored Service entity with identifier: %s is not allowed"
-                    + "with notification rules if it's created through template by reference.",
-                monitoredServiceDTO.getIdentifier()));
+  private boolean validateUpdate(
```
Comment (karansaraswat19) [inline srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/core/services/impl/monitoredService/MonitoredServiceServiceImpl.java:502]: what is isTemplatised here?

**Review by deepak-harness — DISMISSED** [review]
> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/55938)* status: :shipit: complete! ✓

File: srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/core/services/impl/monitoredService/MonitoredServiceServiceImpl.java (hunk 3 above)
Comment (naveen-harness) [reply srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/core/services/impl/monitoredService/MonitoredServiceServiceImpl.java:502]: `isTemplatised` basically means whether the entity is supported in template reconciliation. 
For change source and health source, it's true 
for notification and dependencies, it's false

**Comment by naveen-harness** [comment]
trigger uts

**Comment by naveen-harness** [comment]
trigger uts

**Comment by naveen-harness** [comment]
trigger uts

**Review by karansaraswat19 — APPROVED** [review]

**Comment by naveen-harness** [comment]
trigger ut8

**Review by n00bitax — APPROVED** [review]

**Pull request merged**